### PR TITLE
Add activation support to Tensile

### PR DESCRIPTION
**Pull request by KKyang**
OS: Ubuntu 18.04.5 LTS

Tested ``unit``, ``pre_checkin``, ``extended``, ``integration`` with tox. New tests are added in ``Tensile/Tests/activaiton``.

### CODE REVIEW

**Review by jichangjichang — COMMENTED** [review]
combine with main commit

File: Tensile/ActivationType.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+################################################################################
```
Comment (jichangjichang) [inline Tensile/ActivationType.py:1]: suggest merging into Activation.py

File: Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -1962,6 +1685,9 @@ def initKernel(self, kernel, tPA, tPB ):
     self.numSgprOffsetC = 1
     self.numSgprOffsetA = 1
     self.numSgprOffsetB = 1
+    self.numActivationTypeArgSize = 0 # Will change to 1 if activationType == All
+    self.numActivationArgSize = max(1, int(kernel["ProblemType"]["DestDataType"].numBytes() / 4))
```
Comment (jichangjichang) [inline Tensile/KernelWriterAssembly.py:1689]: .numRegisters()

**Review by jichangjichang — COMMENTED** [review]
Just left some comments for now and review is still ongoing.

File: Tensile/KernelWriterAssembly.py [hunk 3]
```diff
@@ -2086,6 +1812,13 @@ def initKernel(self, kernel, tPA, tPB ):
     self.defineSgpr("Alpha", numSgprAlpha, numSgprAlpha)
     if kernel["ProblemType"]["UseBeta"]:
       self.defineSgpr("Beta", numSgprBeta, numSgprBeta)
+    if ((kernel["ProblemType"]["ActivationType"] != 'none') and (kernel["_GlobalAccumulation"] != 'MultipleBuffer') \
+        and (globalParameters["ActivationNoFuse"] == False)):
```
[the remark this answers is not in the document]
Comment (KKyang) [reply Tensile/KernelWriterAssembly.py:1816]: This is for debug only.

File: Tensile/Activation.py [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+      kStr += reg0FiveInitStr
+      kStr += self.inst("v%s_mul_f16"%pkStr, self.getVgprStr(vgprIdx), vgpr(vgpr0Five), vgpr(vgprtmp), "0.5 * x * (1 + tanh(...))" )
+      if reg0FiveInitStr:
+        self.vgprActivationPool.unregisterTemp(vgpr0Five)
+      self.vgprActivationPool.unregisterTemp(vgprtmp)
+    elif cDataType.isSingle():
+      vgprtmp, needInit = self.vgprActivationPool.registerTemp(1, 'gelu vgpr temp', "activation exp")
+      kStr += self.inst("v_mul_f32", vgpr(vgprtmp), self.getVgprStr(vgprIdx), self.getVgprStr(vgprIdx), "x * x" )
+      kStr += self.inst("v_mul_f32", vgpr(vgprtmp), ActivationMagicNumbers["FloatGeluK1"], vgpr(vgprtmp), "k1 * x * x" )
+      kStr += self.inst("v_add_f32", vgpr(vgprtmp), 1.0, vgpr(vgprtmp), "1 + k1 * x * x" )
```
[the remark this answers is not in the document]
Comment (KKyang) [reply Tensile/Activation.py:274]: This is fixed in newer commit.

File: Tensile/ClientWriter.py [hunk 5]
```diff
@@ -477,6 +481,23 @@ def param(key, value):
...
+            param('activation-enum-args', setting.activationEnum.toEnum())
+        if problemType.activationType == 'all':
+          activationCDataType = problemType.betaType if problemType.activationHPA else \
+                                problemType.dType
+          enumList = ActivationType.getEnumStrList(activationCDataType)
+          wlStr = ""
+          for actType in enumList:
+            wlStr += ActivationType(actType).toEnum() + ","
+          wlStr = wlStr.rstrip(',')
+          param('activation-whitelist', wlStr)
```
Comment (jichangjichang) [inline Tensile/ClientWriter.py:495]: Predicating problem should be determined by individual kernel capability.
but this list is given by caller (user).

**Review by jichangjichang — COMMENTED** [review]
beside the comment, 
enabling tensile debug would encounter Segmentation fault.
ex:
TENSILE_DB=0xFF ../Tensile/bin/Tensile test_activation_hpa_int8.yaml out_folder

File: Tensile/Activation.py [hunk 6]
```diff
@@ -0,0 +1,800 @@
...
+      kStr += self.inst("v_cmp_ge_f64", self.vcc, self.getVgprStr(vgprIdx, 2), self.getSgprStr(activationAlpha, 2), "x >= alpha ?")
+      kStr += self.inst("v_cndmask_b32", self.getVgprStr(vgprIdx), 0, self.getVgprStr(vgprIdx), self.vcc, "set x to 0 if < 0")
+      kStr += self.inst("v_cndmask_b32", self.getVgprStr(vgprIdx+1), 0, self.getVgprStr(vgprIdx+1), self.vcc, "set x to 0 if < 0")
+      kStr += self.inst("v_min_f64", self.getVgprStr(vgprIdx, 2), self.getSgprStr(activationBeta, 2), self.getVgprStr(vgprIdx, 2), "min(x, beta)")
+    elif cDataType.isInt32():
+      kStr += self.inst("v_cmp_ge_i32", self.vcc, self.getVgprStr(vgprIdx), self.getSgprStr(activationAlpha), "x >= alpha ?")
+      kStr += self.inst("v_cndmask_b32", self.getVgprStr(vgprIdx), 0.0, self.getVgprStr(vgprIdx), self.vcc, "set x to 0 if < alpha")
+      kStr += self.inst("v_min_i32", self.getVgprStr(vgprIdx), self.getSgprStr(activationBeta), self.getVgprStr(vgprIdx), "min(x, beta)")
+    return kStr
+  # Internal Use
```
Comment (jichangjichang) [inline Tensile/Activation.py:430]: This way would be more straightforward and easy for reading.
> import math
> import struct
> 
> def double_to_hex(f):
>     return hex(struct.unpack('<Q', struct.pack('<d', f))[0])
> def float_to_hex(f):
>     return hex(struct.unpack('<I', struct.pack('<f', f))[0])
> def f16_to_hex(f):
>     return hex(struct.unpack('<H', struct.pack('<e', f))[0])
>     
> inv_ln2=1/math.log(2)
> print("inverse ln2 ", inv_ln2)
> print(double_to_hex(inv_ln2))
> print(float_to_hex(inv_ln2))
> print(f16_to_hex(inv_ln2))
> print()
> 
> neg_inv_ln2=-1*inv_ln2
> print("negative inverse ln2 ", neg_inv_ln2)
> print(double_to_hex(neg_inv_ln2))
> print(float_to_hex(neg_inv_ln2))
> print(f16_to_hex(neg_inv_ln2))

**Result**
> inverse ln2  1.4426950408889634
> 0x3ff71547652b82fe
> 0x3fb8aa3b
> 0x3dc5
> 
> negative inverse ln2  -1.4426950408889634
> 0xbff71547652b82fe
> 0xbfb8aa3b
> 0xbdc5

File: Tensile/SolutionStructs.py [hunk 7]
```diff
@@ -860,6 +864,11 @@ def __init__(self, config):
       print2("DataType == f16 and HPA == True; setting compute data type to f32")
       self["ComputeDataType"] = DataType('s')
 
+    # Modifying ComputeDataType for I8I8I_BH: if (I8I8I8+HPA), convert it to I8I8I_BH by setting ComputeDataType to i.
```
Comment (jichangjichang) [inline Tensile/SolutionStructs.py:867]: ComputeDataType may use float

**Pull request closed without merging**